Parse command line arguments with Boost

svenstucki opened the pull request:
I've had a go at issue #177. It's a simple change, but there are still some points that are up for discussion.

For example, I don't fully understand each available option yet - maybe you could provide some accurate descriptions for the options with a TODO.

The most important point is how to exit when `--help` is used. In the short term an exception can be used, however it's not very clean.

In the longer term command line parsing should probably be decoupled from the SDL/engine initialization and maybe even merged with config file parsing (btw. Boost has support for that, but might not work together with the current config setup). This would require more rewriting though.

danhedron commented:
I already want to move some initialisation out from the engine init so that we can access the logger in the outermost exception handler. Moving the command line parsing can be done in another PR later that improves initialisation.

danhedron commented on rwgame/RWGame.cpp:
```diff
@@ -48,39 +49,60 @@ RWGame::RWGame(int argc, char* argv[])
+		("height,h",     po::value<size_t>(),      "Game resolution height in pixel")
+		("fullscreen,f",                           "Enable fullscreen mode")
+		("newgame",                                "Start a new game") // TODO
+		("test",                                   "Testmode") // TODO
```
This interacts with `newgame`, it creates some vehicles & weapons instead of starting the game script.

JayFoxRox commented on rwgame/RWGame.cpp:
```diff
@@ -48,39 +49,60 @@ RWGame::RWGame(int argc, char* argv[])
+		("height,h",     po::value<size_t>(),      "Game resolution height in pixel")
+		("fullscreen,f",                           "Enable fullscreen mode")
+		("newgame,n",                              "Directly start a new game")
+		("test,t",                                 "To be used with -n, starts in a test location instead")
```
This works without `-n` iirc.

You'd just use `-t`, `-l` or `-n` to skip the menu (each represents a menu option).
(I've never used `-b` but it's probably also in that group)

svenstucki replied:
No, I've checked that. With `-t` only it shows the main menu. Otherwise I would've figured out that this corresponds to the Test menu entry :).

svenstucki replied:
Actually, I meant "debug" menu entry.

Maybe this can be considered a bug/enhancement though.

JayFoxRox commented:
You should add the new boost dep here so we can get proper travis output https://github.com/rwengine/openrw/blob/master/.travis.yml#L13 :smile: 

Note that I'm not even sure about using boost anymore, I was under the impression this part was a header-only lib. \- I'm currently working on the MSVC port of OpenRW and boost is a giant pita (250MB) dependency. Even stripped to only headers it's like 30MB. We are using it only for filesystem and the unit tests right now.
As the unit tests are optional and filesystem is not actually necessary for MSVC (as it also implements C++17 filesystem already) we could ditch boost as a dependency on Windows asap, and in a couple of months when C++17 has gained some ground. We'd only require a replacement for this and `boost::iequals`.

It's just food for thought though. You are doing a good job here and we can probably use this for a while. *IF* we decide to ditch boost at a later point we can probably still switch to another header-only lib way easier than with our current code :)

(I also recommend joining us on irc #openrw / freenode)

svenstucki commented:
As discussed on IRC - the latest commit should be a good compromise between not feeling too dirty and not overdoing it in a single PR.

The constructor no (should) no longer throw Boost exceptions, but uses a `std::invalid_argument` exception to exit whenever an invalid argument or `--help` is present.

JayFoxRox commented on rwgame/RWGame.cpp:
```diff
@@ -67,14 +68,20 @@ RWGame::RWGame(int argc, char* argv[])
-		// TODO: This is a hack
 		std::cout << desc << std::endl;
-		throw std::runtime_error("Terminate");
+		throw std::invalid_argument("");
```
nit: shouldn't this only happen in the `help` case?
I didn't test yet, but it looks like `./rwgame --help` would cause invalid argument? Is that standard? Here is a little `cat` (meow) experiment:

```
$ cat  --error; echo $?
cat: unrecognized option '--error'
Try 'cat --help' for more information.
```
Returns 1
```
$ cat  --help; echo $?
Usage: cat [OPTION]... [FILE]...
Concatenate FILE(s) to standard output.
...
Full documentation at: <http://www.gnu.org/software/coreutils/cat>
or available locally via: info '(coreutils) cat invocation'
```
Returns 0

svenstucki replied:
Yeah, as I tried to explain, the exception is just a crutch to exit the game from the RWGame constructor.

I've added the unrecognized option text. It will only show the first "wrong" option, but so does e.g. cat. For example:
```
$ cat -a -b
cat: invalid option -- 'a'
```

The only thing that's left is the return code. It will indeed be the same for both a wrong command line argument and `--help`. This is a bit unfortunate, but would be easier to fix after reorganizing the code a bit.

danhedron commented on rwgame/RWGame.cpp:
```diff
@@ -48,39 +49,67 @@ RWGame::RWGame(int argc, char* argv[])
+		("newgame,n",                              "Directly start a new game")
+		("test,t",                                 "Starts a new game in a test location")
+		("load,l",       po::value<std::string>(), "Load save file")
+		("benchmark,b",  po::value<std::string>(), "Run benchmark and store results in file")
```
The benchmark only prints the result at the moment

svenstucki replied:
Ah, no idea why I thought it's storing results in a file. Fixed that.

danhedron commented:
All options seem to work. Looks close to ready imo.

svenstucki commented:
I've rebased to current master and added some small improvements regarding your comments above.

Btw. I don't think Boost has a way to make `--newgame`, `--test`, etc. exclusive (i.e. making having more than one of those arguments an error). The way around this would be to change the options a bit, e.g. use `--run <newgame|test|load|benchmark> [<file>]` (`--run` would default to newgame) instead. It's a bit harder to parse, but certainly doable.

danhedron commented:
I'm keen to get this merged so that #235 can be re-based. There's an issue when passing invalid or ambiguous arguments:
```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::program_options::ambiguous_option> >'
  what():  option '--h' is ambiguous and matches '--height', and '--help'
fish: “./rwgame/rwgame --h” terminated by signal SIGABRT (Abort)
```
or
```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::program_options::invalid_option_value> >'
  what():  the argument ('text') for option '--width' is invalid
fish: “./rwgame/rwgame --width text” terminated by signal SIGABRT (Abort)
```
I think just catching ``boost::program_options::error`` should handle all these other errors descriptively enough. You may want to squash or fix-up some of the smaller commits, but I'll leave that up to you.

Any further tidying up can be done when the initialisation is reorganised.

svenstucki commented:
You're right, I completely forgot to handle other exceptions. Fixed with your recommendation.

I've reduced the number of commits a bit.

danhedron commented:
Great, thanks for doing this. It looks ready for merging to me, do you mind updating the title?

Tested & reviewed.

svenstucki commented:
Nice! Thanks for accepting it.
